luigi/contrib/gcs.py

```diff
@@ -400,29 +398,11 @@ def download(self, path, chunksize=None, chunk_callback=lambda _: False):
             request = self.client.objects().get_media(bucket=bucket, object=obj)
             downloader = http.MediaIoBaseDownload(fp, request, chunksize=chunksize)
 
-            attempts = 0
-            done = False
-            while not done:
-                error = None
-                try:
-                    _, done = downloader.next_chunk()
-                    if chunk_callback(fp):
-                        done = True
-                except errors.HttpError as err:
-                    error = err
-                    if err.resp.status < 500:
-                        raise
-                    logger.warning('Error downloading file, retrying', exc_info=True)
-                except RETRYABLE_ERRORS as err:
-                    logger.warning('Error downloading file, retrying', exc_info=True)
-                    error = err
-
-                if error:
-                    attempts += 1
-                    if attempts >= NUM_RETRIES:
-                        raise error
-                else:
-                    attempts = 0
+            try:
+                _, done = downloader.next_chunk()
+            except errors.HttpError as err:
+                if err.resp.status < 500:
```

Comment:
I don't think this check is in the right place. You will silently swallow 5xx errors. IIUC, you'll want to retry on those.

If you remove the `try .. except` wrapper here, and replace `retry_if_exception_type` with a custom function that checks the exception type and status code, I think that we should be fine.

Reply:
I implemented `custom_retry` function to evaluate exception and status code. Please review.
Thank you for your comment, again!